execution_engine/src/core/runtime/mod.rs

```diff
@@ -1852,36 +1846,123 @@ where
             .ok_or(Error::InvalidContractVersion(contract_version_key))?;
 
         // Get contract data
-        let contract = match self.context.read_gs(&contract_hash.into())? {
+        let contract_key = contract_hash.into();
+        let contract = match self.context.read_gs(&contract_key)? {
             Some(StoredValue::Contract(contract)) => contract,
             Some(_) => {
-                return Err(Error::FunctionNotFound(format!(
-                    "Value at {:?} is not a contract",
-                    contract_package_hash
-                )));
+                return Err(Error::InvalidContract(contract_hash));
             }
-            None => return Err(Error::KeyNotFound(key)),
+            None => return Err(Error::KeyNotFound(contract_key)),
         };
 
         let entry_point = contract
             .entry_point(&entry_point_name)
             .cloned()
             .ok_or_else(|| Error::NoSuchMethod(entry_point_name.to_owned()))?;
 
+        self.call_contract_checked(contract_package, contract_hash, contract, entry_point, args)
+    }
+
+    /// Calls contract if caller has access, and args match entry point definition
+    fn call_contract_checked(
+        &mut self,
+        contract_package: ContractPackage,
+        contract_hash: ContractHash,
+        contract: Contract,
+        entry_point: EntryPoint,
+        args: RuntimeArgs,
+    ) -> Result<CLValue, Error> {
+        // if public, allowed
+        // if not public, restricted to user group access
         self.validate_entry_point_access(&contract_package, entry_point.access())?;
 
-        for (expected, found) in entry_point
-            .args()
-            .iter()
-            .map(|a| a.cl_type())
-            .cloned()
-            .zip(args.to_values().into_iter().map(|v| v.cl_type()).cloned())
+        // This will skip arguments check for system contracts only. This code should be removed on
+        // next major version bump. Argument checks for system contract is still done during
+        // execution of a system contract.
+        if !self
+            .protocol_data()
+            .system_contracts()
+            .contains(&contract_hash)
```

Comment:
@henrytill this line skips entry point arguments check for system contracts only. We can't upgrade entry points at the next minor/patch version bump without changing the upgrade procedure, so I figured the safest thing to do is just to not check the arguments for these until the next major version bump. This PR is ready for the next major bump (which will install fixed entrypoints) and has regression tests in place to verify it.